lstm_word_segmentation/word_segmenter_cnn.py

```diff
@@ -0,0 +1,596 @@
+        probs /= probs.sum(axis=-1, keepdims=True)
+
+        return probs.astype(dtype)
+    
+    def segment_arbitrary_line(self, input_line):
+        """
+        This function uses the CNN model to segment an unsegmented line and compare it to ICU and deepcut.
+        Args:
+            input_line: the string that needs to be segmented. It is supposed to be unsegmented
+        """
+        line = Line(input_line, "unsegmented")
+        grapheme_clusters_in_line = len(line.char_brkpoints) - 1
+
+        # Using CNN model to segment the input line
+        if self.embedding_type == "codepoints":
+            x_data = []
+            for i in range(len(line.unsegmented)):
+                x_data.append(CodePoint(line.unsegmented[i], self.codepoint_dic))
+        else:
+            x_data = []
+            for i in range(grapheme_clusters_in_line):
+                char_start = line.char_brkpoints[i]
+                char_finish = line.char_brkpoints[i + 1]
+                curr_char = line.unsegmented[char_start: char_finish]
+                x_data.append(GraphemeCluster(curr_char, self.graph_clust_dic, self.letters_dic))
+        y_hat = Bies(input_bies=self._manual_predict(x_data), input_type="mat")
+
+        # Making a pretty version of the output of the CNN, where bars show the boundaries of words
+        y_hat_pretty = ""
+        if self.embedding_type == "codepoints":
+            for i in range(len(line.unsegmented)):
+                if y_hat.str[i] in ['b', 's']:
+                    y_hat_pretty += "|"
+                y_hat_pretty += line.unsegmented[i]
+            y_hat_pretty += "|"
+        else:
+            y_hat_pretty = ""
+            for i in range(grapheme_clusters_in_line):
+                char_start = line.char_brkpoints[i]
+                char_finish = line.char_brkpoints[i + 1]
+                curr_char = line.unsegmented[char_start: char_finish]
+                if y_hat.str[i] in ['b', 's']:
+                    y_hat_pretty += "|"
+                y_hat_pretty += curr_char
+            y_hat_pretty += "|"
+
+        return y_hat_pretty
+
+    def _get_trainable_data(self, input_line):
+        """
+        Given a segmented line, generates a list of input data (with respect to the embedding type) and a n*4 np array
+        that represents BIES where n is the length of the unsegmented line.
+        Args:
+            input_line: the unsegmented line
+        """
+        # Finding word breakpoints
+        # Note that it is possible that input is segmented manually instead of icu. However, for both cases we set that
+        # input_type equal to "icu_segmented" because that doesn't affect performance of this function. This way we
+        # won't need unnecessary if/else for "man_segmented" and "icu_segmented" throughout rest of this function.
+        line = Line(input_line, "icu_segmented")
+
+        # x_data and y_data will be code point based if self.embedding_type is codepoints
```

Comment:
Here and in other places, you have "code point". I think they all need to be "codepoint".